python/neml2/pyzag/interface.py

```diff
@@ -0,0 +1,287 @@
+        )
+
+        self.neml2_sep_char = neml2_sep_char
+        self.python_sep_char = python_sep_char
+        self._setup_parameters(exclude_parameters)
+
+        self.state_axis = state_axis
+        self.forces_axis = forces_axis
+        self.residual_axis = residual_axis
+        self.old_prefix = old_prefix
+
+        self._check_model()
+
+    def _setup_parameters(self, exclude_parameters):
+        """Mirror parameters of the neml model with torch.nn.Parameter
+
+        Args:
+            exclude_parameters (list of str): neml parameters to exclude
+        """
+        self.parameter_name_map = {}
+        for neml_name, neml_param in self.model.named_parameters().items():
+            if neml_name in exclude_parameters:
+                continue
+            neml_param.requires_grad_(True)
+            rename = neml_name.replace(self.neml2_sep_char, self.python_sep_char)
+            self.parameter_name_map[rename] = neml_name
+            self.register_parameter(rename, torch.nn.Parameter(neml_param.torch()))
+
+    def _update_parameter_values(self):
+        """Copy over new parameter values"""
+        self.model.set_parameters(
+            {
+                self.parameter_name_map[n]: Tensor(
+                    p,
+                    self.model.get_parameter(self.parameter_name_map[n])
+                    .tensor()
+                    .batch.dim(),
+                )
+                for n, p in self.named_parameters()
+            }
+        )
+
+    def _check_model(self):
+        """Simple consistency checks, could be a debug check but we only call this once"""
+        should_axes = [
+            self.state_axis,
+            self.old_prefix + self.state_axis,
+            self.forces_axis,
+            self.old_prefix + self.forces_axis,
+        ]
+        if self.model.input_axis().nsubaxis() != len(should_axes):
+            raise ValueError(
+                "Wrapped NEML2 model should only have 4 subaxes on the input axis"
+            )
+        for axis in should_axes:
+            if not self.model.input_axis().has_subaxis(AA(axis)):
+                raise ValueError("Wrapped NEML2 model missing input subaxis %s" % axis)
+
+        # Output axis should just have the residual
+        if self.model.output_axis().nsubaxis() != 1:
+            raise ValueError(
+                "Wrapped NEML2 model should only have 1 subaxes on the output axis"
+            )
+
+        if not self.model.output_axis().has_subaxis(AA(self.residual_axis)):
+            raise ValueError(
+                "Wrapped NEML2 model is missing required output subaxis %s"
+                % self.residual_axis
+            )
+
+        # And all the variables on state should match the variables in the residual
+        for name in self.model.input_axis().subaxis(self.state_axis).variable_names():
+            self.model.output_axis().subaxis(self.residual_axis).has_variable(AA(name))
```

Comment:
```suggestion
        # And all the variables on state should match the variables in the residual
        if self.model.input_axis().subaxis(self.state_axis) != self.model.output_axis().subaxis(self.residual_axis):
            raise ValueError("...")
```

Reply:
The C++ side has this `!=` operator. If the python binding is missing just add it.

Reply:
Cool, but I would like to be able to tell the user which variable(s) is missing.

Reply:
Oh then you need your original code. But I think you forgot to raise error.

Reply:
That is a good point.

Reply:
Actually what I would do in this case is that I'll use the != operator to do the check, and if it errors print out the axes in the error message and let the user scrutinize.

Your approach might be throwing the error too early upon first missing variable...